File: isoft_angola_tax_compliance/isoft_angola_tax_compliance/report/invalid_nif/invalid_nif.py
# Copyright (c) 2026, ISOFT and contributors
# For license information, please see license.txt
"""Customers whose NIF does not match the Angolan format.

Grouped so the two very different problems do not look the same:

  Wrong Customer Type   the NIF is valid, the type is not -- fix the type
  Invalid Format        the NIF itself is wrong -- fix the number

`Invoices` counts submitted Sales Invoices already issued against that customer,
so the ones that matter most sort to the top.
"""

import frappe
from frappe import _

from isoft_angola_tax_compliance.nif import PATTERNS, check, get_settings


def execute(filters=None):
	filters = frappe._dict(filters or {})
	return get_columns(), get_data(filters)


def get_columns():
	return [
		{"label": _("Customer"), "fieldname": "customer", "fieldtype": "Link",
		 "options": "Customer", "width": 240},
		{"label": _("Customer Name"), "fieldname": "customer_name", "fieldtype": "Data", "width": 220},
		{"label": _("Type"), "fieldname": "customer_type", "fieldtype": "Data", "width": 90},
		{"label": _("NIF"), "fieldname": "tax_id", "fieldtype": "Data", "width": 150},
		{"label": _("Problem"), "fieldname": "problem", "fieldtype": "Data", "width": 160},
		{"label": _("Suggested Type"), "fieldname": "suggested_type", "fieldtype": "Data", "width": 130},
		{"label": _("Invoices"), "fieldname": "invoices", "fieldtype": "Int", "width": 90},
		{"label": _("Detail"), "fieldname": "detail", "fieldtype": "Data", "width": 420},
	]


def get_data(filters):
	settings = get_settings()

	conditions = ["1=1"]
	values = {}
	if not filters.get("include_disabled"):
		conditions.append("IFNULL(c.disabled, 0) = 0")
	if filters.get("customer_type"):
		conditions.append("c.customer_type = %(customer_type)s")
		values["customer_type"] = filters.customer_type

	customers = frappe.db.sql(
		"""
		SELECT c.name, c.customer_name, c.customer_type, IFNULL(c.tax_id, '') AS tax_id
		FROM `tabCustomer` c
		WHERE {conditions}
		ORDER BY c.name
		""".format(conditions=" AND ".join(conditions)),
		values,
		as_dict=True,
	)

	bad = []
	for row in customers:
		ok, message = check(row.tax_id, row.customer_type, settings)
		if ok:
			continue

		suggested = ""
		for other, pattern in PATTERNS.items():
			if other != row.customer_type and pattern.match((row.tax_id or "").strip()):
				suggested = other
				break

		bad.append(
			{
				"customer": row.name,
				"customer_name": row.customer_name,
				"customer_type": row.customer_type,
				"tax_id": row.tax_id,
				"problem": _("Wrong Customer Type") if suggested else _("Invalid Format"),
				"suggested_type": suggested,
				"detail": frappe.utils.strip_html(message or ""),
			}
		)

	if not bad:
		return []

	# One query for the invoice counts rather than one per customer.
	names = [r["customer"] for r in bad]
	counts = dict(
		frappe.db.sql(
			"""SELECT customer, COUNT(*) FROM `tabSales Invoice`
			   WHERE docstatus = 1 AND customer IN %(names)s GROUP BY customer""",
			{"names": names},
		)
		or []
	)
	for row in bad:
		row["invoices"] = counts.get(row["customer"], 0)

	# Most-invoiced first: those are the ones already reaching AGT.
	bad.sort(key=lambda r: (-r["invoices"], r["customer"]))
	return bad

File: isoft_angola_tax_compliance/isoft_angola_tax_compliance/report/invalid_nif/invalid_nif.js
// Copyright (c) 2026, ISOFT and contributors
// For license information, please see license.txt

frappe.query_reports["Invalid NIF"] = {
	filters: [
		{
			fieldname: "customer_type",
			label: __("Customer Type"),
			fieldtype: "Select",
			options: ["", "Individual", "Company"],
		},
		{
			fieldname: "include_disabled",
			label: __("Include Disabled Customers"),
			fieldtype: "Check",
			default: 0,
		},
	],
	onload(report) {
		report.page.add_inner_button(__("Fix Customer Types"), () => atc_nif_fix_types(report));
	},
	formatter(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		if (column.fieldname === "problem" && data) {
			const colour = data.problem === __("Wrong Customer Type") ? "orange" : "red";
			value = `<span style="color: var(--text-on-${colour}, ${colour})">${value}</span>`;
		}
		return value;
	},
};


// Bulk-correct Customer Type where the NIF is unambiguously valid for the other
// type. Malformed numbers are never touched -- they stay in the report.
function atc_nif_fix_types(report) {
	const include_disabled = report.get_filter_value("include_disabled") ? 1 : 0;

	frappe.call({
		method: "isoft_angola_tax_compliance.nif.get_type_fix_preview",
		args: { include_disabled },
		freeze: true,
		freeze_message: __("Checking which customers can be corrected…"),
		callback(r) {
			const preview = r && r.message;
			if (!preview) return;

			if (!preview.total) {
				frappe.msgprint({
					title: __("Nothing to correct"),
					indicator: "blue",
					message: __(
						"No customer has a NIF that is valid for the other type. " +
							"Anything left in this report needs the number itself fixed."
					),
				});
				return;
			}

			const directions = Object.keys(preview.by_direction)
				.map((k) => `<li><b>${frappe.utils.escape_html(k)}</b>: ${preview.by_direction[k]}</li>`)
				.join("");

			const rows = (preview.sample || [])
				.map(
					(f) => `<tr>
						<td>${frappe.utils.escape_html(f.customer)}</td>
						<td>${frappe.utils.escape_html(f.tax_id)}</td>
						<td>${frappe.utils.escape_html(f.from_type)}</td>
						<td><b>${frappe.utils.escape_html(f.to_type)}</b></td>
					</tr>`
				)
				.join("");

			const remaining = preview.total - (preview.sample || []).length;

			const dialog = new frappe.ui.Dialog({
				title: __("Fix Customer Types"),
				size: "large",
				fields: [
					{
						fieldtype: "HTML",
						fieldname: "preview",
						options: `
							<p>${__("<b>{0}</b> customer(s) have a NIF that is valid for the other type.", [
								preview.total,
							])}</p>
							<ul class="small">${directions}</ul>
							<p class="text-muted small">
								${__(
									"Only the Customer Type is changed, and only where the NIF format leaves no doubt. Customers whose NIF is malformed are not touched."
								)}
							</p>
							<div style="max-height: 320px; overflow: auto;">
							<table class="table table-bordered table-condensed small">
								<thead><tr>
									<th>${__("Customer")}</th><th>${__("NIF")}</th>
									<th>${__("From")}</th><th>${__("To")}</th>
								</tr></thead>
								<tbody>${rows}</tbody>
							</table>
							</div>
							${remaining > 0 ? `<p class="text-muted small">${__("… and {0} more.", [remaining])}</p>` : ""}
						`,
					},
				],
				primary_action_label: __("Correct {0} Customer(s)", [preview.total]),
				primary_action() {
					dialog.hide();
					frappe.call({
						method: "isoft_angola_tax_compliance.nif.run_type_fix",
						args: { include_disabled },
						freeze: true,
						freeze_message: __("Correcting…"),
						callback(res) {
							const updated = (res && res.message && res.message.updated) || 0;
							frappe.show_alert(
								{ message: __("{0} customer(s) corrected.", [updated]), indicator: "green" },
								7
							);
							report.refresh();
						},
					});
				},
			});

			dialog.show();
		},
	});
}
